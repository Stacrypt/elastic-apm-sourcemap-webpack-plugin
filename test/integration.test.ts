import webpack from 'webpack';
import path from 'path';
import mockFetch from 'jest-fetch-mock';
import fetch from 'node-fetch';
import ElasticAPMSourceMapPlugin, { Config } from '../src/elastic-apm-sourcemap-webpack-plugin';

jest.mock('node-fetch', () => mockFetch);

const getWebpackConfig = (pluginConfig: Config): webpack.Configuration => ({
  entry: path.resolve(__dirname, './entry.js'),
  mode: 'production',
  devtool: 'source-map',
  plugins: [new ElasticAPMSourceMapPlugin(pluginConfig)]
});

beforeEach(() => {
  fetch.resetMocks();
});

test('ok', cb => {
  fetch.mockResponse(JSON.stringify('ok'));

  webpack(
    getWebpackConfig({
      serviceName: 'mock-service',
      serviceVersion: 'mock-version',
      publicPath: '/mock-folder',
      serverURL: 'mock-url'
    }),
    (err, stats) => {
      if (err) {
        return cb(err);
      }

      if (stats.hasErrors()) {
        return cb(stats.toJson().errors);
      }

      expect(fetch.mock.calls.length).toEqual(1);
      expect(fetch.mock.calls[0][0]).toEqual('mock-url');
      expect(fetch.mock.calls[0][1].method).toEqual('POST');

      // TODO: check body

      cb();
    }
  );
});

test('failed', cb => {
  fetch.mockReject('failed');

  webpack(
    getWebpackConfig({
      serviceName: 'mock-service',
      serviceVersion: 'mock-version',
      publicPath: '/mock-folder',
      serverURL: 'mock-url'
    }),
    err => {
      expect(fetch.mock.calls.length).toEqual(1);
      expect(err).toEqual('failed');
      setTimeout(() => {
        cb();
      }, 100);
    }
  );
});

test('400', cb => {
  fetch.mockResponses(['failed', { status: 400 }]);

  webpack(
    getWebpackConfig({
      serviceName: 'mock-service',
      serviceVersion: 'mock-version',
      publicPath: '/mock-folder',
      serverURL: 'mock-url'
    }),
    err => {
      expect(fetch.mock.calls.length).toEqual(1);
      expect(err).toBeInstanceOf(Error);
      setTimeout(() => {
        cb();
      }, 100);
    }
  );
});

test('with secret', cb => {
  fetch.mockResponse(JSON.stringify('ok'));

  webpack(
    getWebpackConfig({
      serviceName: 'mock-service',
      serviceVersion: 'mock-version',
      publicPath: '/mock-folder',
      serverURL: 'mock-url',
      secret: 'mock-secret'
    }),
    () => {
      expect(fetch.mock.calls[0][1].headers).toEqual({ Authorization: 'Bearer mock-secret' });
      cb();
    }
  );
});
