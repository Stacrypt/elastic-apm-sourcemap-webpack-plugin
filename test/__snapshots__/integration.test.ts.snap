// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fail to reach the server 1`] = `
Array [
  Array [
    "starting uploading sourcemaps with configs: {\\"logLevel\\":\\"warn\\",\\"ignoreErrors\\":false,\\"serviceName\\":\\"mock-service\\",\\"serviceVersion\\":\\"mock-version\\",\\"publicPath\\":\\"/mock-folder\\",\\"serverURL\\":\\"mock-url\\"}.",
  ],
  Array [
    "uploading main.js.map to Elastic APM with bundle_filepath: /mock-folder/main.js.",
  ],
]
`;

exports[`fail to reach the server 2`] = `
Array [
  Array [
    "failed",
  ],
]
`;

exports[`send to the server successfully 1`] = `
Array [
  Array [
    "starting uploading sourcemaps with configs: {\\"logLevel\\":\\"warn\\",\\"ignoreErrors\\":false,\\"serviceName\\":\\"mock-service\\",\\"serviceVersion\\":\\"mock-version\\",\\"publicPath\\":\\"/mock-folder\\",\\"serverURL\\":\\"mock-url\\"}.",
  ],
  Array [
    "uploading main.js.map to Elastic APM with bundle_filepath: /mock-folder/main.js.",
  ],
  Array [
    "uploaded main.js.map.",
  ],
  Array [
    "finished uploading sourcemaps.",
  ],
]
`;

exports[`send to the server successfully 2`] = `
"--FIXED-BOUNDARY
Content-Disposition: form-data; name=\\"sourcemap\\"; filename=\\"main.js.map\\"
Content-Type: application/json

{\\"version\\":3,\\"sources\\":[\\"webpack://elastic-apm-sourcemap-webpack-plugin/./test/entry.js\\"],\\"names\\":[\\"console\\",\\"log\\"],\\"mappings\\":\\"AAAAA,QAAQC,IAAI\\",\\"file\\":\\"main.js\\",\\"sourcesContent\\":[\\"console.log('hello');\\\\n\\"],\\"sourceRoot\\":\\"\\"}
--FIXED-BOUNDARY
Content-Disposition: form-data; name=\\"service_version\\"

mock-version
--FIXED-BOUNDARY
Content-Disposition: form-data; name=\\"bundle_filepath\\"

/mock-folder/main.js
--FIXED-BOUNDARY
Content-Disposition: form-data; name=\\"service_name\\"

mock-service
--FIXED-BOUNDARY--
"
`;
